helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -136,11 +143,14 @@ private Optional<AssignableNode> getNodeWithHighestPoints(AssignableReplica repl
     }).collect(Collectors.toList());
 
     if (candidateNodes.isEmpty()) {
+      enableFullLoggingForReplica(replica);
       optimalAssignment.recordAssignmentFailure(replica,
           Maps.transformValues(hardConstraintFailures, this::convertFailureReasons));
       return Optional.empty();
     }
 
+    removeReplicaFromFullLogging(replica);
```

Comment:
Same here, we can have paired log to understand we exit it.

Reply:
Yes,  Updated the PR.